### ISSUE TITLE
Grpc.Tools attempts to run 64bit protoc even on 32-bit linux system.

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#18543]
```
"/var/local/git/grpc/test/distrib/csharp/DistribTest.sln" (default target) (1) ->
"/var/local/git/grpc/test/distrib/csharp/DistribTest/DistribTest.csproj" (default target) (2) ->
(_Protobuf_CoreCompile target) -> 
  /var/local/git/grpc/test/distrib/csharp/packages/Grpc.Tools.1.19.0/build/_protobuf/Google.Protobuf.Tools.targets(263,5): error MSB6006: "/var/local/git/grpc/test/distrib/csharp/packages/Grpc.Tools.1.19.0/tools/linux_x64/protoc" exited with code 255. [/var/local/git/grpc/test/distrib/csharp/DistribTest/DistribTest.csproj]

    3 Warning(s)
    1 Error(s)

Time Elapsed 00:00:05.68
+ FAILED=true
+ '[' '' '!=' '' ']'
+ docker rm -f build_and_run_docker_ad5750a7-2495-4db7-9228-75d6a234defc
build_and_run_docker_ad5750a7-2495-4db7-9228-75d6a234defc
+ '[' true '!=' '' ']'
+ exit 1
FAILED: distribtest.csharp_linux_x86_jessie [ret=1, pid=187578, time=28.0sec]
```

**Comment by jtattermusch** [comment]
CC @kkm000  this error happens when running inside an 32bit docker container on a 64bit machine, 
which doesn't seem to be handled well by the platform detection code in 
https://github.com/grpc/grpc/blob/4ef6a274271406f961f2fc9cc639ae204d4d128e/src/csharp/Grpc.Tools/Common.cs#L71

**Comment by jtattermusch** [comment]
This is a bit related to https://github.com/grpc/grpc/pull/17437

**Comment by kkm000** [comment]
Is this mono or dotnet?

**Comment by kkm000** [comment]
Is 32-bit container even _possible_ on a 64-bit host? I thought a container always uses the host's kernel. This is a container, not a VM?

**Comment by jtattermusch** [comment]
It's mono. 

A bit more background:
Well it's a 32bit linux libraries running on a 64bit linux. (you can get a docker image that has a 32bit installation of debian - obviously the 64bit protoc is not going to work there because all the system C libraries are missing and only 32bit libs are "installed" - a 32bit build protoc will work just fine, but we need to make sure to select it).

**Comment by kkm000** [comment]
A 64-bit kernel can run 32-bit processes, so a "32-bit container" is just an image with 32-bit ld.so and executables. Just like running 32-bit process on a normal 64-bit machine. The os is still 64 bit, but there is no 64-bit ld.so. Weird. Wound the `linux32` command help?

```
kkm@yupana:~$ docker run --rm -it i386/debian:9
Unable to find image 'i386/debian:9' locally
9: Pulling from i386/debian
3cf1c3e5c1c3: Pull complete
Digest: sha256:657c9a7aa7735b0f0273562c2f55e5dc1a3783ff70dbdb1a3c415d978f24c84d
Status: Downloaded newer image for i386/debian:9
root@bcb4d6cff153:/# uname -a
Linux bcb4d6cff153 4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 x86_64 GNU/Linux
root@bcb4d6cff153:/# linux32 uname -a
Linux bcb4d6cff153 4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 i686 GNU/Linux
root@bcb4d6cff153:/# 
```

**Comment by jtattermusch** [comment]
In the meantime I came up with this: 
https://github.com/grpc/grpc/pull/18467/files#diff-3af54eacd0b2cf68b28672cd2651034dR91

**Comment by kkm000** [comment]
Oh, and there is also `RuntimeInformation.ProcessArchitecture`. I do not know which one is which. I can only *hope* the Environment.Is32BitProcess (the only one existing before dotnet) mirrors this one.

Maybe we should always use the current process' architecture, for both Mono and dotnet, what do you think? The idea is If we are running inside an N-bit MSBuild, then the OS is capable of running N-bit processes. Should be true for any N, unless maybe it's a Mac. These can easily beat any math an logic, in my experience, so it's better to verify...

Is this Docker or lxc (or Google's own)? I remember there was a difference; AFAICR, lxc is capable of somehow virtualizing the container, essentially turning it into a little VM. I do not understand this mechanism. But all 32-bit Docker images invariably package `linux32`. What happens if you try running msbuild command with `linux32 msbuild` in the test? Or is it just not worth checking?

**Comment by kkm000** [comment]
Looks like `linux32` does not have much effect at all on my machine. I thought it was a real thing...

```cs
$ cat bitness.cs
using System.Runtime.InteropServices;
class _ {
  static void Main() {
    System.Console.WriteLine("OS={0} Process={1} Env.Is64Os={2} Env.Is64Proc={3}",
      RuntimeInformation.OSArchitecture,
      RuntimeInformation.ProcessArchitecture,
      System.Environment.Is64BitOperatingSystem,
      System.Environment.Is64BitProcess);
} }
```
**Mono**
```
$ file $(realpath $(which mono))
/usr/bin/mono-sgen: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, <snip long build id>
$ csc bitness.cs
Microsoft (R) Visual C# Compiler version 2.8.2.62916 (2ad4aabc)
Copyright (C) Microsoft Corporation. All rights reserved.

$ mono bitness.exe
OS=X64 Process=X64 Env.Is64Os=True Env.Is64Proc=True
$ linux32 mono bitness.exe
OS=X64 Process=X64 Env.Is64Os=True Env.Is64Proc=True
$ rm bitness.exe
```

**dotnet**
```
$ file $(realpath $(which dotnet))
/usr/share/dotnet/dotnet: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, <snip>
$ dotnet new console -o .
. . .
$ rm Program.cs
$ dotnet run --runtime linux-x86
OS=X64 Process=X64 Env.Is64Os=True Env.Is64Proc=True
$ linux32 dotnet run --runtime linux-x86
OS=X64 Process=X64 Env.Is64Os=True Env.Is64Proc=True
$ dotnet run --runtime linux-x64
OS=X64 Process=X64 Env.Is64Os=True Env.Is64Proc=True
$ linux32 dotnet run --runtime linux-x64
OS=X64 Process=X64 Env.Is64Os=True Env.Is64Proc=True
```
Not as simple as I thought it would be. Only uname is easily fooled. Sheesh.

Looks like I need to do that in a container. Likely tomorrow, need to do something now.

**Comment by kkm000** [comment]
Ah, nailed it for mono! I did not even know it had this mkbundle tool!
```
$ mkbundle --fetch-target mono-5.18.0-debian-9-x86
$ mkbundle --cross mono-5.18.0-debian-9-x86 --deps bitness.exe
From: /home/kkm/.mono/targets/mono-5.18.0-debian-9-x86
Using runtime: /home/kkm/.mono/targets/mono-5.18.0-debian-9-x86/bin/mono
     Assembly: /home/kkm/work/mklver/bitness.exe
     Assembly: /home/kkm/.mono/targets/mono-5.18.0-debian-9-x86/lib/mono/4.5/mscorlib.dll
     Assembly: /home/kkm/.mono/targets/mono-5.18.0-debian-9-x86/lib/mono/4.5/I18N.West.dll
     Assembly: /home/kkm/.mono/targets/mono-5.18.0-debian-9-x86/lib/mono/4.5/I18N.dll
systemconfig: /home/kkm/.mono/targets/mono-5.18.0-debian-9-x86/etc/mono/config
machineconfig: /home/kkm/.mono/targets/mono-5.18.0-debian-9-x86/etc/mono/4.5/machine.config
Generated a.out
$ file a.out
a.out: ELF 32-bit LSB shared object, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux.so.2, <snip>
$ ./a.out
OS=X64 Process=X86 Env.Is64Os=True Env.Is64Proc=False
```
and in the same binary in a 32-bit container
```
$ docker container ls -l
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                          PORTS               NAMES
709202a16a9f        i386/debian:9       "bash"              2 minutes ago       Exited (0) About a minute ago                       optimistic_hugle
$ docker cp a.out optimistic_hugle:/root/bitness
$ docker commit optimistic_hugle kkm/mono32
$ docker run --rm -it kkm/mono32
root@e861bfe55d34:/# cd root
root@e861bfe55d34:~# ./bitness
OS=X64 Process=X86 Env.Is64Os=True Env.Is64Proc=False
```
Looks like we can trust `System.Environment.Is64BitProcess`. I have only a 64 bit runtime currently, but if mono does it, then I have no doubt that dotnet will.

optimistic_hugle, huh! Not a dictionary word, but sounds indeed optimistic :)

**Comment by jtattermusch** [comment]
Fixed by #18467.
